rtl/eclrtl/rtlformat.hpp

```diff
@@ -278,6 +285,32 @@ class ECLRTL_API CommonJsonWriter : implements IXmlWriterExt, public CInterface
     bool needDelimiter;
 };
 
+class ECLRTL_API CommonJsonObjectWriter : public CommonJsonWriter
+{
+public:
+    CommonJsonObjectWriter(unsigned _flags, unsigned initialIndent,  IXmlStreamFlusher *_flusher) : CommonJsonWriter(_flags, initialIndent, _flusher)
+    {
+        outputBeginRoot();
+    }
+    ~CommonJsonObjectWriter()
+    {
+        if (!final)
+            outputEndRoot();
+    }
+    virtual const char * finalStr() override
+    {
+        if (!final)
+        {
+            final=true;
+            outputEndRoot();
+        }
+        return str();
+    }
```

Comment:
a new virtual finalStr() vs str() seems a bit odd/a bit of a shame. And may be confusing that finalStr() == str() if finalStr() called first.

It could be argued that an other writers should logically have an equivalent that performs final closing.

Perhaps it would be better to introduce new openRoot(), closeRoot() to IXmlWriterExt, which are null ops. for most except this impl. - and then the only current use contact (sockfile) can call open/close in addition to normal str(), instead of finalStr()

Reply:
@jakesmith yeah, but for streams the current way is less obtrusive.  Destroying the object would finalize the output without new calls.  Only in the more rare case where we are building and extracting the string do we need to say we want it finalized.  So do we want to minimize the complexity for most use cases, or make it more explicit using new root based calls for all?

Reply:
I agree do not want to introduce extra calls for symmetry into common use cases that don't actually require them.

Still seems odd that (a = str()) != (b = finalStr()) == (c = str()) though.

I'm not sure of all the use cases of the current IXmlWriterExt::str() ...
are there calls to it that are not 'final' ? i.e. is it called to get content of partial progress?
Or are all use cases actually only when writing done?

If so, how about making CommonJsonObjectWriter::str() itself performing the final end root functionality?

Reply:
@afishbeck - yes, I think that would be a bit clearer.

>Worried that either str() could be used in cases where the object wasn't final

Looking at all the use cases (there are 9 in community), they all look as if they are 'final', IOW - if str() were to finalize anything they needed, there'd be no change in behaviour.
Although I think it would also be prudent to change the method name if that was the case from 'str()' to imply that that it finalized too, like your 'finalstr()', i.e. only have finalStr() - and ideally assert if anything was used after that before a reset.

Anyway, probably beyond the scope of this change, I'd go for a new 'finalize()' method, leaving str() as it was in this PR.